cli/index.d.ts

```diff
@@ -269,6 +269,9 @@ export abstract class Transform {
   /** Lists all files in a directory. */
   listFiles(dirname: string, baseDir: string): (string[] | null) | Promise<string[] | null>;
 
+  /** Called when file was read, before parsing the content. */
+  afterRead?(file: { sourceText: string, sourcePath: string }): Promise<void>;
```

Comment:
Wondering: Would it perhaps make sense to change the signature to `afterRead?(sourceText: string, sourcePath: string)`, and let the function return the potentially modified source text? Would ensure that `sourcePath` is not accidentally changed, and perhaps aid eventual C-like bindings if there isn't a by-reference object.

Reply:
Good point. Also thinking about it

Reply:
However `sourceText` in this case also will be unmodified which may be a useful.

Reply:
Perhaps something like this?
```ts
  afterRead?(filePath: string, contentText: string): Promise<string | null>;
```
and in `cli/index.js` use this:
```js
    const file = { sourceText, sourcePath };
    let resOrError = await applyTransform("afterRead", sourcePath, sourceText);
    if (resOrError) {
       if (resOrError instanceof Error) return prepareResult(error);
       file.sourceText = resOrError;
    }
    return file;
 }
```

Reply:
Can't this become...?

```ts
try {
  sourceText = await applyTransform("afterRead", sourcePath, sourceText);
} catch (e) {
  return prepareResult(error);
}
```

Reply:
`applyTransform` already wrap to try / catch. But doesn't propagate return value from `transform` hook
```js
try {
          let start = stats.begin();
          stats.transformCount++;
          await transform[name](...args);
          stats.transformTime += stats.end(start);
        } catch (e) {
          return e;
        }
```

Reply:
MaxGraey, what necessary steps do we need to complete the feature?

Reply:
Well it seems to finish this also required add some significant refactoring to transform's tests. They are failing due to we now have 2 async methods but check globals which not proper way anymore. I'll add PR for that soon

Reply:
ok, maybe we can split implementation? 1) first simple sync version 2) then I can create feature request for async version?

Reply:
The main problem is we currently have a situation when parse exist between read files is messed:
```ts
// Parse entry files
  {
    let code = await parseBacklog();
    if (code) return code;
  }

  // Call afterParse transform hook
  {
    let error = await applyTransform("afterParse", program.parser);
    if (error) return prepareResult(error);
  }

  // Parse additional files, if any
  {
    let code = await parseBacklog();
    if (code) return code;
  }
```
`parseBacklog` internally call `getFile` in parallel way. See [this part](https://github.com/AssemblyScript/assemblyscript/blob/main/cli/index.js#L680-L696). This lead to fail in transform tests and this will be a problem with async transforms inside `getFile`. I don't know how to best fix this. @dcodeIO any ideas?

Reply:
Hmm, though one. Looks like the hook could instead be inserted once all the async operations are completed, maintaining ordering?

Reply:
@dcode why do you need suck big refactoring for current task? Or refactoring here is needed anyway, so you want to make them both once?